sympy/simplify/cse_diff.py

```diff
@@ -0,0 +1,190 @@
+"""Module for differentiation using CSE."""
+
+from collections import Counter
+
+from sympy.core.containers import Tuple
+from sympy.core.singleton import S
+from sympy.core.symbol import Symbol
+from sympy.core.traversal import postorder_traversal
+from sympy.matrices.immutable import ImmutableDenseMatrix
+from sympy.simplify.cse_main import CseExpr, cse
+from sympy.utilities.iterables import numbered_symbols
+from sympy.physics.mechanics import dynamicsymbols
+
+
+def _forward_jacobian(
+    expr: ImmutableDenseMatrix,
+    wrt: ImmutableDenseMatrix,
```

Comment:
This seems a bit restrictive. It would be better to accept anything that quacks like a matrix, i.e. has rows and columns with entries that are sympy expressions and `wrt` should be any iterable of sympy variables (symbols or functions).